### PR TITLE
UCT/CUDA/GTEST:  Testing MD memory type interface

### PR DESCRIPTION
**Pull request by bureddy**

### CODE REVIEW

**Comment by bureddy** [comment]
@yosefe @alex-mikheev please review

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5597/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5599/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5600/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
bot:mlnx:retest

**Comment by bureddy** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5608/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5609/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5610/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_md.c [hunk 1]
```diff
@@ -152,8 +152,15 @@ static ucs_status_t uct_ib_md_query(uct_md_h uct_md, uct_md_attr_t *md_attr)
                              UCT_MD_FLAG_NEED_MEMH |
                              UCT_MD_FLAG_NEED_RKEY |
                              UCT_MD_FLAG_ADVISE;
-    md_attr->cap.reg_mem_types = UCS_BIT(UCT_MD_MEM_TYPE_HOST) |
-                                 UCS_BIT(UCT_MD_MEM_TYPE_CUDA);
+    md_attr->cap.reg_mem_types = UCS_BIT(UCT_MD_MEM_TYPE_HOST);
+
+#if HAVE_CUDA
+    /* check if GDR driver is loaded */
+    if (!access("/sys/kernel/mm/memory_peers/nv_mem/version", F_OK)) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: it's there a better way to check this?

File: test/gtest/uct/test_md.cc [hunk 2]
```diff
@@ -106,6 +113,64 @@ void test_md::check_caps(uint64_t flags, const std::string& name)
...
+void test_md::check_memory(void *address, size_t pos, int expect, int mem_type)
+{
+    if (mem_type == UCT_MD_MEM_TYPE_HOST) {
+        EXPECT_EQ(expect, *((char*)address + pos));
+    } else if (mem_type == UCT_MD_MEM_TYPE_CUDA) {
+#if HAVE_CUDA
+        void *temp;
+        cudaError_t ret;
+
+        temp = malloc(1);
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc]: why not check the entire buffer?

File: test/gtest/uct/test_md.cc [hunk 3]
```diff
@@ -199,78 +264,115 @@ UCS_TEST_P(test_md, alloc) {
...
+    ASSERT_UCS_OK(status);
+
+    if (md_attr.cap.mem_type == UCT_MD_MEM_TYPE_HOST) {
+        UCS_TEST_SKIP_R("MD owns only host memory");
+    }
+
+    alloc_memory(&address, 1024, 0xBB, md_attr.cap.mem_type);
+
+    ret = uct_md_is_mem_type_owned(pd(), address, 1024);
+    EXPECT_GT(ret, 0);
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc]: EXPECT_TRUE

File: test/gtest/uct/test_md.cc [hunk 4]
```diff
@@ -199,78 +264,115 @@ UCS_TEST_P(test_md, alloc) {
...
+                continue;
+            }
 
-        address = malloc(size);
-        ASSERT_TRUE(address != NULL);
+            alloc_memory(&address, size, 0xBB, mem_type);
 
-        memset(address, 0xBB, size);
+            status = uct_md_mem_reg(pd(), address, size, UCT_MD_MEM_ACCESS_ALL,
+                    &memh);
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc]: indentation

Comment (bureddy) [reply test/gtest/uct/test_md.cc]: Fixed

File: test/gtest/uct/test_md.cc (hunk 3 above)
Comment (bureddy) [reply test/gtest/uct/test_md.cc]: Fixed

File: test/gtest/uct/test_md.cc (hunk 2 above)
Comment (bureddy) [reply test/gtest/uct/test_md.cc]: Just wanted to keep the same  checking as host buffer case where it is checking last byte content

File: src/uct/ib/base/ib_md.c (hunk 1 above)
Comment (bureddy) [reply src/uct/ib/base/ib_md.c]: another way is to parse /proc/modules and search for "nv_peer_mem". But, i think this is better

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5615/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5616/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5617/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

File: test/gtest/uct/test_md.cc (hunk 2 above)
Comment (yosefe) [reply test/gtest/uct/test_md.cc]: so let's check all buffer in both cases
for example use ucs::fill_random() to generate random data

File: src/uct/ib/base/ib_md.c (hunk 1 above)
Comment (yosefe) [reply src/uct/ib/base/ib_md.c]: ok

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5620/ for details (Mellanox internal link).

File: test/gtest/uct/test_md.cc (hunk 2 above)
Comment (bureddy) [reply test/gtest/uct/test_md.cc]: fixed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5631/ for details (Mellanox internal link).

**Pull request merged**